### PR TITLE
Enclave data sealing for private keys

### PR DESCRIPTION
**Pull request by pgarg66**
This change adds data sealing/unsealing functionality to SGX signing enclave.

### CODE REVIEW

File: src/sgx/signing/signing.edl [hunk 1]
```diff
@@ -9,14 +9,40 @@ enclave {
     include "signing_internal.h"
 
    trusted {
+      public sgx_status_t init_remote_attestation(
```
Comment (rob-solana) [inline src/sgx/signing/signing.edl]: documentation for this stuff?

**Comment by pgarg66** [comment]
HI @rob-solana , do you have any additional feedback?

**Comment by pgarg66** [comment]
https://github.com/solana-labs/solana/issues/1429

File: src/sgx/signing/signing_public.h [hunk 2]
```diff
@@ -15,6 +16,12 @@ typedef struct ed25519_context {
   uint8_t public_key[ED25519_PUB_KEY_LEN];
 } ed25519_context_t;
 
+typedef struct history_entry {
+  uint32_t num_hashes;
+  uint32_t optional_input_hash;
```
Comment (rob-solana) [inline src/sgx/signing/signing_public.h]: these aren't large enough to accommodate actual hashes...

File: src/sgx/signing/signing_public.h [hunk 3]
```diff
@@ -26,7 +33,29 @@ typedef struct ed25519_context {
...
+                              uint32_t* datalen,
+                              uint8_t* data);
+
+/* This function reinitializes the enclave using sealed data.
+ */
+sgx_status_t init_ed25519_from_data(ed25519_context_t* pctxt,
+                                    uint32_t datalen,
+                                    uint8_t* data,
+                                    uint32_t update_lockout_params,
+                                    uint32_t lockout_period,
```
Comment (rob-solana) [inline src/sgx/signing/signing_public.h]: lockout period shouldn't be re-initialize-able, should be part of initial setup

Comment (pgarg66) [reply src/sgx/signing/signing_public.h]: 👍 
I think APIs will be updated based on final design of the service. I am leaving them as is for now, as the signatures are still in flux.

Per your suggestion, I've added code for rollback protection. That'll be useful to protect against reinit.

**Pull request merged**